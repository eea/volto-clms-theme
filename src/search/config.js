import { mergeConfig } from '@eeacms/search';
import { getClientProxyAddress, getSearchThumbUrl } from './utils';
import vocabs from './vocabulary';
import facets from './facets';
import views from './views';

const clmsConfig = {
  title: 'CLMS Main',
  ...views,
};

export default function install(config) {
  const envConfig = process.env.RAZZLE_ENV_CONFIG
    ? JSON.parse(process.env.RAZZLE_ENV_CONFIG)
    : clmsConfig;

  const pjson = require('@eeacms/volto-clms-theme/../package.json');

  envConfig.app_name = pjson.name;
  envConfig.app_version = pjson.version;

  config.searchui.clmsSearchTechnicalLibrary = {
    ...mergeConfig(envConfig, config.searchui.globalsearchbase),
    elastic_index: '_es/clmsSearchTechnicalLibrary',
    index_name: 'copernicus_searchui',
    host: process.env.RAZZLE_ES_PROXY_ADDR || 'http://localhost:3000',
    ...vocabs,
  };

  const { clmsSearchTechnicalLibrary } = config.searchui;

  clmsSearchTechnicalLibrary.permanentFilters.push({
    term: {
      cluster_name: 'copernicus_land',
    },
  });

  clmsSearchTechnicalLibrary.facets = facets;

  clmsSearchTechnicalLibrary.resultItemModel = {
    factory: 'ResultModel',
    urlField: 'about',
    titleField: 'title',
    metatypeField: 'objectProvides',
    descriptionField: 'description',
    tagsField: 'topic',
    issuedField: 'issued',
    getThumbnailUrl: 'getSearchThumbUrl',
    getIconUrl: 'getGlobalsearchIconUrl',
  };

  clmsSearchTechnicalLibrary.sortOptions = [
    {
      name: 'Relevance',
      value: '',
      direction: '',
    },
    {
      name: 'Newest',
      value: 'issued.date',
      direction: 'desc',
    },
    {
      name: 'Oldest',
      value: 'issued.date',
      direction: 'asc',
    },
    {
      name: 'Version ↑',
      value: 'version.keyword',
      direction: 'asc',
    },
    {
      name: 'Version ↓',
      value: 'version.keyword',
      direction: 'desc',
    },
  ];
  config.resolve.getSearchThumbUrl = getSearchThumbUrl();

  if (typeof window !== 'undefined') {
    config.searchui.clmsSearchTechnicalLibrary.host =
      process.env.RAZZLE_ES_PROXY_ADDR || getClientProxyAddress();
  }

  return config;
}
