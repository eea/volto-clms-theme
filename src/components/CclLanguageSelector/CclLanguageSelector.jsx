/**
 * Language selector component.
 * @module components/LanguageSelector/LanguageSelector
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import cookie from 'react-cookie';
import { useSelector, useDispatch } from 'react-redux';
import cx from 'classnames';
import { find } from 'lodash';
import { updateIntl } from 'react-intl-redux';
import config from '@plone/volto/registry';
import { flattenToAppURL } from '@plone/volto/helpers';
import CclModal from '@eeacms/volto-clms-theme/components/CclModal/CclModal';
import { FormattedMessage } from 'react-intl';

import { Helmet, langmap } from '@plone/volto/helpers'; // volto 13.0.1 upgrade

let locales = {};

if (config.settings) {
  config.settings.supportedLanguages.forEach((lang) => {
    import('~/../locales/' + lang + '.json').then((locale) => {
      locales = { ...locales, [lang]: locale.default };
    });
  });
}

function Capitalize(str) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}

function CclLanguageSelector(props) {
  const dispatch = useDispatch();
  const currentLang = useSelector((state) => state.intl.locale);
  const translations = useSelector(
    (state) => state.content.data?.['@components']?.translations?.items,
  );

  function changeLanguage(language) {
    cookie.save('I18N_LANGUAGE', language, {
      expires: new Date((2 ** 31 - 1) * 1000),
      path: '/',
    });

    dispatch(
      updateIntl({
        locale: language,
        messages: locales[language],
      }),
    );
  }

  return config.settings.isMultilingual ? (
    <div className="ccl-header-lang">
      <CclModal
        trigger={
          <div className="header-lang-icon">
            <i className="ccl-icon-language"></i>
            <span className="header-lang-code">
              {currentLang.toUpperCase()}
            </span>
          </div>
        }
        size="fullscreen"
      >
        <div className="ccl-container">
          <div className="modal-language-header">
            <div className="modal-language-title">
              <i className="ccl-icon-language"></i>
              <h3 className="modal-title">
                <FormattedMessage
                  id="Select your language"
                  defaultMessage="Select your language"
                />
              </h3>
            </div>
          </div>
          <div className="modal-language-body">
            <div className="language-list">
              {config.settings.supportedLanguages.map((lang) => {
                const translation = find(translations, { language: lang });
                // console.log(langmap[lang]);
                return (
                  <div
                    key={lang}
                    className={cx(lang, 'language-item', {
                      'language-item-selected': lang === currentLang,
                    })}
                  >
                    <span className="language-link" lang-code={lang}>
                      {lang !== currentLang ? (
                        <Link
                          to={
                            translation
                              ? flattenToAppURL(translation['@id'])
                              : `/${lang}`
                          }
                          title={langmap[lang].nativeName}
                          onClick={() => {
                            props.onClickAction();
                            changeLanguage(lang);
                          }}
                          key={`language-selector-${lang}`}
                        >
                          {Capitalize(langmap[lang].nativeName)}
                        </Link>
                      ) : (
                        Capitalize(langmap[lang].nativeName)
                      )}
                    </span>
                  </div>
                );
              })}
            </div>
          </div>
        </div>
      </CclModal>

      <div className="header-lang-text">
        <CclModal
          trigger={<span>{Capitalize(langmap[currentLang].nativeName)}</span>}
          size="fullscreen"
        >
          <div className="ccl-container">
            <div className="modal-language-header">
              <div className="modal-language-title">
                <i className="ccl-icon-language"></i>
                <h3 className="modal-title">
                  <FormattedMessage
                    id="Select your language"
                    defaultMessage="Select your language"
                  />
                </h3>
              </div>
            </div>
            <div className="modal-language-body">
              <div className="language-list">
                {config.settings.supportedLanguages.map((lang) => {
                  const translation = find(translations, { language: lang });
                  return (
                    <div
                      key={lang}
                      className={cx(lang, 'language-item', {
                        'language-item-selected': lang === currentLang,
                      })}
                    >
                      <span className="language-link" lang-code={lang}>
                        {lang !== currentLang ? (
                          <Link
                            to={
                              translation
                                ? flattenToAppURL(translation['@id'])
                                : `/${lang}`
                            }
                            title={langmap[lang].nativeName}
                            onClick={() => {
                              props.onClickAction();
                              changeLanguage(lang);
                            }}
                            key={`language-selector-${lang}`}
                          >
                            {Capitalize(langmap[lang].nativeName)}
                          </Link>
                        ) : (
                          Capitalize(langmap[lang].nativeName)
                        )}
                      </span>
                    </div>
                  );
                })}
              </div>
            </div>
          </div>
        </CclModal>
      </div>
    </div>
  ) : (
    <Helmet>
      <html lang={config.settings.defaultLanguage} />
    </Helmet>
  );
}

CclLanguageSelector.propTypes = {
  onClickAction: PropTypes.func,
};

CclLanguageSelector.defaultProps = {
  onClickAction: () => {},
};

export default CclLanguageSelector;
