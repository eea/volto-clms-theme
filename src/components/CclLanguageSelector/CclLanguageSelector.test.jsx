import config from '@plone/volto/registry';
import React from 'react';
import { Provider } from 'react-intl-redux';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';

import '@testing-library/jest-dom';

import CclLanguageSelector from './CclLanguageSelector';

jest.mock('@plone/volto/helpers/Loadable/Loadable');
beforeAll(async () => {
  await require('@plone/volto/helpers/Loadable/Loadable').__setLoadables();
});

beforeAll(() => {
  config.settings.supportedLanguages = ['es'];
});

const mockStore = configureStore();

describe('CclLanguageSelector', () => {
  it('Renders no multilingual', () => {
    config.settings.isMultilingual = false;
    const store = mockStore({
      intl: {
        locale: 'en',
        messages: {},
      },
      content: {
        data: {
          '@components': {
            translations: {
              items: [{ language: 'es', '@id': '/es' }],
            },
          },
        },
      },
    });
    const component = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <CclLanguageSelector
            onClickAction={() => {
              return 'test';
            }}
          ></CclLanguageSelector>
          ,
        </MemoryRouter>
      </Provider>,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('Renders Multilingual', () => {
    config.settings.isMultilingual = true;
    const store = mockStore({
      intl: {
        locale: 'en',
        messages: {},
      },
      content: {
        data: {
          '@components': {
            translations: {
              items: [{ language: 'es', '@id': '/es' }],
            },
          },
        },
      },
    });
    const component = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <CclLanguageSelector
            onClickAction={() => {
              return 'test';
            }}
          ></CclLanguageSelector>
          ,
        </MemoryRouter>
      </Provider>,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('Test click event', async () => {
    config.settings.isMultilingual = true;
    const store = mockStore({
      intl: {
        locale: 'en',
        messages: {},
      },
      content: {
        data: {
          '@components': {
            translations: {
              items: [{ language: 'es', '@id': '/es' }],
            },
          },
        },
      },
    });
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <CclLanguageSelector />
        </MemoryRouter>
      </Provider>,
    );

    await waitFor(() => {
      const headerDiv = container.querySelector('div.header-lang-icon');
      expect(headerDiv).toBeInTheDocument();
      fireEvent.click(headerDiv);

      const mobileLangLink = screen.getByLabelText('Switch to espa√±ol');
      expect(mobileLangLink).toBeInTheDocument();

      const deskSel = container.querySelector('.header-lang-text span');
      expect(deskSel).toBeInTheDocument();
      fireEvent.click(deskSel);
    });
  });
});
