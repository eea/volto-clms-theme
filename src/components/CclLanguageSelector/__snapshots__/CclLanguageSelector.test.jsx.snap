// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CclLanguageSelector Renders Multilingual 1`] = `
Array [
  <div
    className="ccl-header-lang"
  >
    <div
      className="header-lang-icon"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <i
        className="ccl-icon-language"
      />
      <span
        className="header-lang-code"
      >
        en
      </span>
    </div>
    <div
      className="header-lang-text"
    >
      <span
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        English
      </span>
    </div>
  </div>,
  ",",
]
`;

exports[`CclLanguageSelector Renders no multilingual 1`] = `","`;
